File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    answer = []
    terms_obj = {}
    year = 12 * 28
    now_year, now_month, now_day = list(map(int, today.split(".")))
    today_days = now_year * year + now_month * 28 + now_day
    
    for term in terms:
        type, time = term.split(" ")
        terms_obj[type] = int(time)
    
    for index, item in enumerate(privacies):
        date, type = item.split(" ")
        item_year, item_month, item_day = list(map(int, date.split(".")))
        target_date = item_year * year + (item_month + terms_obj[type]) * 28 + (item_day - 1)
        
        if today_days > target_date:
            answer.append(index + 1)
            
    return answer

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.js
function solution(today, terms, privacies) {
    const answer = [];
    
    // today를 Array형태로 변환
    today = today.split(".").map((item) => +item);
    // 유효기간 확인을 위한 객체 생성
    const expireRef = {};
    terms.forEach((item) => {
        const targetItem = item.split(" ");
        expireRef[targetItem[0]] = +targetItem[1];
    })
    
    // privacies을 순회하면서 유효기간이 지났는지 확인
    privacies.forEach((item, index) => {
        item = item.split(" ");
        const checkDate = item[0].split(".").map((item) => +item);
        const checkType = item[1];
        
        checkDate[1] += expireRef[checkType];
            // console.log({checkDate, checkType}, 444)
        if (checkDate[1] > 12) {
            checkDate[0] += parseInt((checkDate[1] - 1) / 12);
            checkDate[1] = (checkDate[1] - 1) % 12 + 1;
        }

        if (checkDate[0] < today[0]) {
            answer.push(index + 1);
        } else if (checkDate[0] === today[0]) {
            if (checkDate[1] < today[1]) {
                answer.push(index + 1);
            } else if (checkDate[1] === today[1]) {
                if (checkDate[2] <= today[2]) {
                    answer.push(index + 1);
                }
            }
        }
    })
    return answer;
}